lectures/oop_intro.md

```diff
@@ -244,9 +251,513 @@ x
+
+```{code-cell} ipython
+x = 2
+y = 3
+
+import numpy as np
+
+%whos
+```
+
+### The Global Namespace
+
+```{index} single: Python; Namespace (Global)
+```
+
+Python documentation often makes reference to the "global namespace".
+
+The global namespace is *the namespace of the module currently being executed*.
+
+For example, suppose that we start the interpreter and begin making assignments .
+
+We are now working in the module `__main__`, and hence the namespace for `__main__` is the global namespace.
+
+Next, we import a module called `amodule`
+
+```{code-block} python3
+:class: no-execute
+
+import amodule
+```
+
+At this point, the interpreter creates a namespace for the module `amodule` and starts executing commands in the module.
+
+While this occurs, the namespace `amodule.__dict__` is the global namespace.
+
+Once execution of the module finishes, the interpreter returns to the module from where the import statement was made.
+
+In this case it's `__main__`, so the namespace of `__main__` again becomes the global namespace.
+
+### Local Namespaces
+
+```{index} single: Python; Namespace (Local)
+```
+
+Important fact: When we call a function, the interpreter creates a *local namespace* for that function, and registers the variables in that namespace.
+
+The reason for this will be explained in just a moment.
+
+Variables in the local namespace are called *local variables*.
+
+After the function returns, the namespace is deallocated and lost.
+
+While the function is executing, we can view the contents of the local namespace with `locals()`.
+
+For example, consider
+
+```{code-cell} python3
+def f(x):
+    a = 2
+    print(locals())
+    return a * x
+```
+
+Now let's call the function
+
+```{code-cell} python3
+f(1)
+```
+
+You can see the local namespace of `f` before it is destroyed.
+
+### The `__builtins__` Namespace
+
+```{index} single: Python; Namespace (__builtins__)
+```
+
+We have been using various built-in functions, such as `max(), dir(), str(), list(), len(), range(), type()`, etc.
+
+How does access to these names work?
+
+* These definitions are stored in a module called `__builtin__`.
+* They have their own namespace called `__builtins__`.
+
+```{code-cell} python3
+dir()[0:10]
+```
+
+```{code-cell} python3
+dir(__builtins__)[0:10]
+```
+
+We can access elements of the namespace as follows
+
+```{code-cell} python3
+__builtins__.max
+```
+
+But `__builtins__` is special, because we can always access them directly as well
+
+```{code-cell} python3
+max
+```
+
+```{code-cell} python3
+__builtins__.max == max
+```
+
+The next section explains how this works ...
+
+### Name Resolution
+
+```{index} single: Python; Namespace (Resolution)
+```
+
+Namespaces are great because they help us organize variable names.
+
+(Type `import this` at the prompt and look at the last item that's printed)
+
+However, we do need to understand how the Python interpreter works with multiple namespaces .
```

Comment:
It might be nice to say a sentence about why knowing this is useful? 

Something like:

Understanding the flow of execution is important so you know which variables are in scope and how to operate on them when writing and debugging programs.